### DISCUSSION TITLE
Possible to share workqueues across processes (not just across threads)?

### DISCUSSION THREAD

**Discussion by JonKohler** [axboe/liburing#513] [Q&A]
We've got an application that instantiates a completely separate process per user, with each process having its own threads/co-routines. These co-routines do many things, including IO using read/readv and write/writev. Sometimes there are hundreds of users connected to a single host, such that there would be hundreds of these processes handling IO with their co-routines.

We're looking at implementing io_uring in this process, where we'd roughly have an io_uring per thread, that co-routines could submit/complete IO against. That's all well and good. We'd like use IORING_SETUP_ATTACH_WQ such that _**within**_ the process would share a workqueue, and perhaps even use sq poll.

As a further optimization, is it possible to share the kernel backend via IORING_SETUP_ATTACH_WQ, such that the io_uring instances _**between**_ these processes bring up could share a workqueue / SQ poller?

**Comment by axboe** [comment]
You can't share the workers across a process, only threads. The reason for that is that the helper needs to present the same mm/credentials/etc to safely handle async operations, and if you have multiple processes involved, then that cannot be done in a sane manner.

**Reply by tofes** [reply]
Is it possible to use multiple io_uring (i.e. one ring per thread)? e.g. One thread does the read and another thread does the write using two separate io_uring? Or does the kernel only expects one io_uring for all IOs per process?

**Reply by axboe** [reply]
The kernel doesn't care, you can have multiple rings per thread and use one for file IO and one for network, etc.
